File: android/app/src/main/java/dev/bridgey/android/ClipboardPayload.kt
package dev.bridgey.android

import org.json.JSONObject

internal const val MAX_CLIPBOARD_CONTENT_BYTES = 32 * 1024

internal data class RichClipboardContent(val text: String, val html: String) {
    fun encode(): ByteArray = JSONObject()
        .put("version", 1)
        .put("text", text)
        .put("html", html)
        .toString()
        .toByteArray(Charsets.UTF_8)

    companion object {
        fun create(text: String, html: String?): RichClipboardContent? {
            val rich = html?.takeIf(String::isNotBlank) ?: return null
            if (text.isEmpty() || text.toByteArray().size + rich.toByteArray().size > MAX_CLIPBOARD_CONTENT_BYTES) return null
            return RichClipboardContent(text, rich)
        }

        fun decode(bytes: ByteArray): RichClipboardContent? = runCatching {
            val objectValue = JSONObject(bytes.toString(Charsets.UTF_8))
            if (objectValue.optInt("version") != 1 || objectValue.opt("text") !is String || objectValue.opt("html") !is String) return null
            create(objectValue.getString("text"), objectValue.getString("html"))
        }.getOrNull()
    }
}

internal fun clipboardTextFits(text: String): Boolean =
    text.isNotEmpty() && text.toByteArray(Charsets.UTF_8).size <= MAX_CLIPBOARD_CONTENT_BYTES

File: macos/Sources/BridgeyMac/ClipboardPayload.swift
import Foundation

let maximumClipboardContentBytes = 32 * 1024

struct RichClipboardContent: Codable, Equatable {
    let version: Int
    let text: String
    let html: String

    init?(text: String, html: String?) {
        guard let html, !text.isEmpty, !html.isEmpty,
              text.utf8.count + html.utf8.count <= maximumClipboardContentBytes else { return nil }
        version = 1
        self.text = text
        self.html = html
    }

    func validated() -> RichClipboardContent? {
        guard version == 1 else { return nil }
        return RichClipboardContent(text: text, html: html)
    }
}

func clipboardTextFits(_ text: String) -> Bool {
    !text.isEmpty && text.utf8.count <= maximumClipboardContentBytes
}
